operators/src/pro/ml/xgboost.rs

```diff
@@ -0,0 +1,694 @@
+        let predicted_grid = crate::util::spawn_blocking(move || {
+            // extract the actual data from the tiles
+            let pixels: Vec<_> = bot
+                .into_iter()
+                .flat_map(|band| -> Result<Vec<f32>, XGBoostModuleError> {
+                    let mat_tile = band.into_materialized_tile();
+
+                    let res = mat_tile
+                        .grid_array
+                        .inner_grid
+                        .data
+                        .into_iter()
+                        .map(num_traits::AsPrimitive::as_)
+                        .collect::<Vec<f32>>();
+
+                    Ok(res)
+                })
+                .flatten()
+                .collect();
+
+            let model_content =
+                std::fs::read_to_string(model).context(self::error::ModelFileParsing)?;
+
+            process_tile(
+                &pixels,
+                &pool,
+                model_content.as_bytes(),
+                grid_shape,
+                n_bands as usize,
+                ndv,
+            )
+        })
+        .await??;
+
+        let rt: BaseTile<GridOrEmpty<GridShape<[usize; 2]>, f32>> =
+            RasterTile2D::new_with_properties(
+                time,
+                tile_position,
+                global_geo_transform,
+                predicted_grid.into(),
+                props.clone(),
+            );
+
+        Ok(rt)
+    }
+}
+
+fn process_tile(
+    bands_of_tile: &Vec<f32>,
+    pool: &ThreadPool,
+    model_file: &[u8],
+    grid_shape: GridShape<[usize; 2]>,
+    n_bands: usize,
+    nan_val: f32,
+) -> Result<geoengine_datatypes::raster::Grid<GridShape<[usize; 2]>, f32>, XGBoostModuleError> {
+    pool.install(|| {
+        // to get one row of data means taking (n_pixels * n_bands) elements
+        let chunk_size = grid_shape.number_of_elements() * n_bands;
+
+        let res: Vec<_> = bands_of_tile
+            .par_chunks(chunk_size)
+            .map(|elem| {
+                // get xgboost style matrices
+                let xg_matrix = DMatrix::from_col_major_f32(
+                    elem,
+                    mem::size_of::<f32>() * n_bands,
+                    mem::size_of::<f32>(),
+                    grid_shape.number_of_elements(),
+                    n_bands,
+                    -1, // TODO: add this to settings.toml: # of threads for xgboost to use
+                    nan_val,
+                )
+                .context(self::error::CreateDMatrix)?;
+
+                let mut out_dim: u64 = 0;
+
+                let bst =
+                    Booster::load_buffer(model_file).context(self::error::LoadBoosterFromFile)?;
+
+                // measure time for prediction
+                let predictions = bst.predict_from_dmat(
+                    &xg_matrix,
+                    &[grid_shape.number_of_elements() as u64, n_bands as u64],
+                    &mut out_dim,
+                );
+
+                match predictions {
```

Comment:
use `map_err` instead of the match block

Reply:
https://github.com/geo-engine/geoengine/pull/639/commits/72aa6c7ba5661768aee327f268897e18182aa408

Is there a general rule for when to use `match` and when `map_err`?

Reply:
if you only convert error types use `map_err` and `.context`. if you do more than that use `match`